### PR TITLE
Initial commit of System.Security.Principal and Principal.Windows

### PR DESCRIPTION
**Pull request by MattGal**
Initial commit of System.Security.Principal and System.Security.Principal.Windows

### CODE REVIEW

**Comment by MattGal** [comment]
Note that while I ported all relevant test collateral, there is very little for Principal.Windows and none for Principal.  We should create more tests for this package.

File: src/System.Security.Principal/src/Var.cs [hunk 1]
```diff
@@ -0,0 +1,6 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+internal static class var { } // Enforce explicit declaration of variable types.
```
Comment (nguerrera) [inline src/System.Security.Principal/src/Var.cs]: Remove this file.

File: src/Common/src/Interop/Windows/NtDll/Interop.NtStatus.cs [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+        // Error codes from ntstatus.h
+        internal const uint STATUS_SUCCESS = 0x00000000;
+        internal const uint STATUS_SOME_NOT_MAPPED = 0x00000107;
+        internal const uint STATUS_NO_MEMORY = 0xC0000017;
+        internal const uint STATUS_OBJECT_NAME_NOT_FOUND = 0xC0000034;
+        internal const uint STATUS_NONE_MAPPED = 0xC0000073;
+        internal const uint STATUS_INSUFFICIENT_RESOURCES = 0xC000009A;
+        internal const uint STATUS_ACCESS_DENIED = 0xC0000022;
+    }
+}
```
Comment (stephentoub) [inline src/Common/src/Interop/Windows/NtDll/Interop.NtStatus.cs]: Nit: some of the files are missing newlines at the end.

File: src/Common/src/Interop/Windows/mincore/Interop.IsEqualDomainSid.cs [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+
+using System;
+using System.Runtime.InteropServices;
+
+internal static partial class Interop
+{
+    internal static partial class mincore
+    {
+        [DllImport(Interop.Libraries.SecurityBase, EntryPoint = "EqualDomainSid", SetLastError = true, CharSet = CharSet.Unicode)]
+        internal static extern int IsEqualDomainSid(byte[] sid1, byte[] sid2, out bool result);
```
Comment (stephentoub) [inline src/Common/src/Interop/Windows/mincore/Interop.IsEqualDomainSid.cs]: Why was this renamed?

File: src/Common/src/Interop/Windows/secur32/Interop.LSAStructs.cs [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+        {
+            Debug.Assert(buffer == null || (stringBytes >= 0 && (ulong)stringBytes <= buffer.ByteLength),
+                            "buffer == null || (stringBytes >= 0 && stringBytes <= buffer.ByteLength)");
+
+            this.Length = (ushort)stringBytes;
+            this.MaxLength = (ushort)buffer.ByteLength;
+
+            // Marshaling with a SafePointer does not work correctly, so unfortunately we need to extract
+            // the raw handle here.
+            this.Buffer = buffer.DangerousGetHandle();
```
Comment (stephentoub) [inline src/Common/src/Interop/Windows/secur32/Interop.LSAStructs.cs]: Shouldn't we be DangerousAddRef'ing and DangerousReleasing as well, before we access DangerousGetHandle and then once we're done with the handle?

File: src/System.Security.Principal.Windows/System.Security.Principal.Windows.sln [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+# Visual Studio 14
+VisualStudioVersion = 14.0.22823.1
+MinimumVisualStudioVersion = 10.0.40219.1
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "System.Security.Principal.Windows", "src\System.Security.Principal.Windows.csproj", "{F9E9894E-2513-4085-9046-311AD49D8AE6}"
+EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "System.Security.Principal.Windows.Tests", "tests\System.Security.Principal.Windows.Tests.csproj", "{6C36F3AC-54A1-4021-9F5D-CDEFF7347277}"
+EndProject
+Global
+	GlobalSection(SolutionConfigurationPlatforms) = preSolution
+		Debug|Any CPU = Debug|Any CPU
```
Comment (stephentoub) [inline src/System.Security.Principal.Windows/System.Security.Principal.Windows.sln]: We should rename these to be Windows_Debug and Windows_Release, both in the .sln and in the .csproj.  This project is Windows only.  We've been using Debug/Release when those configurations work on any OS, and prefixed with OS_Debug/Release when they're specific to an OS.

File: src/System.Security.Principal.Windows/src/Microsoft/Win32/SafeHandles/SafeSecurityHandles.cs [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+    internal abstract class SafeHandleZeroOrMinusOneIsInvalid : SafeHandle
+    {
+        protected SafeHandleZeroOrMinusOneIsInvalid(bool ownsHandle) : base(IntPtr.Zero, ownsHandle)
+        {
+        }
+
+        public override bool IsInvalid
+        {
+            get
+            { return handle == IntPtr.Zero || handle == new IntPtr(-1); }
```
Comment (stephentoub) [inline src/System.Security.Principal.Windows/src/Microsoft/Win32/SafeHandles/SafeSecurityHandles.cs]: Nit: formatting

File: src/System.Security.Principal.Windows/src/Microsoft/Win32/SafeHandles/SafeSecurityHandles.cs [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+
+        // 0 is an Invalid Handle
+        internal SafeLsaMemoryHandle(IntPtr handle) : base(true)
+        {
+            SetHandle(handle);
+        }
+
+        internal static SafeLsaMemoryHandle InvalidHandle
+        {
+            get { return new SafeLsaMemoryHandle(IntPtr.Zero); }
```
Comment (stephentoub) [inline src/System.Security.Principal.Windows/src/Microsoft/Win32/SafeHandles/SafeSecurityHandles.cs]: Should such an invalid handle be cached rather than allocating one each time the property is accessed?

File: src/System.Security.Principal.Windows/src/Microsoft/Win32/SafeHandles/SafeSecurityHandles.cs [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+
+        // 0 is an Invalid Handle
+        internal SafeLsaPolicyHandle(IntPtr handle) : base(true)
+        {
+            SetHandle(handle);
+        }
+
+        internal static SafeLsaPolicyHandle InvalidHandle
+        {
+            get { return new SafeLsaPolicyHandle(IntPtr.Zero); }
```
Comment (stephentoub) [inline src/System.Security.Principal.Windows/src/Microsoft/Win32/SafeHandles/SafeSecurityHandles.cs]: Ditto (I won't comment on this if I see it again).

File: src/System.Security.Principal.Windows/src/System.Security.Principal.Windows.csproj [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
+    <NuGetTargetFrameworkMoniker>ASP.NetCore, version=v5.0</NuGetTargetFrameworkMoniker>
+  </PropertyGroup>
+  <!-- Default configurations to help VS understand the options -->
+  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Debug|AnyCPU' ">
+  </PropertyGroup>
+  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release|AnyCPU' ">
+  </PropertyGroup>
+  <ItemGroup>
+    <ProjectReference Include="..\..\System.Security.Principal\src\System.Security.Principal.csproj">
```
Comment (stephentoub) [inline src/System.Security.Principal.Windows/src/System.Security.Principal.Windows.csproj]: Do we want this to be a NuGet package reference instead of using a P2P reference?  I'm happy with the latter if that's a policy change we're making, but that's not what we've normally done in corefx, and we should be consistent with such things across the repo.

File: src/System.Security.Principal.Windows/src/System/Security/Principal/IRCollection.cs [hunk 10]
```diff
@@ -0,0 +1,440 @@
...
+{
+    public class IdentityReferenceCollection : ICollection<IdentityReference>
+    {
+        #region Private members
+
+        //
+        // Container enumerated by this collection
+        //
+
+        private List<IdentityReference> _Identities;
```
Comment (stephentoub) [inline src/System.Security.Principal.Windows/src/System/Security/Principal/IRCollection.cs]: readonly?

File: src/System.Security.Principal.Windows/src/System/Security/Principal/IRCollection.cs [hunk 11]
```diff
@@ -0,0 +1,440 @@
...
+
+        public bool Remove(IdentityReference identity)
+        {
+            if (identity == null)
+            {
+                throw new ArgumentNullException("identity");
+            }
+            Contract.EndContractBlock();
+
+            if (Contains(identity))
```
Comment (stephentoub) [inline src/System.Security.Principal.Windows/src/System/Security/Principal/IRCollection.cs]: Why not just replace all of this with ```return _Identities.Remove(identity);```?

File: src/System.Security.Principal.Windows/src/System/Security/Principal/IdentityNotMappedException.cs [hunk 12]
```diff
@@ -0,0 +1,43 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace System.Security.Principal
+{
+    public sealed class IdentityNotMappedException : Exception
+    {
+        private IdentityReferenceCollection _unmappedIdentities;
```
Comment (stephentoub) [inline src/System.Security.Principal.Windows/src/System/Security/Principal/IdentityNotMappedException.cs]: readonly?

File: src/System.Security.Principal.Windows/src/System/Security/Principal/NTAccount.cs [hunk 13]
```diff
@@ -0,0 +1,393 @@
...
+using System.Globalization;
+using System.Runtime.InteropServices;
+
+namespace System.Security.Principal
+{
+    public sealed class NTAccount : IdentityReference
+    {
+        #region Private members
+
+        private readonly string _Name;
```
Comment (stephentoub) [inline src/System.Security.Principal.Windows/src/System/Security/Principal/NTAccount.cs]: Some of these private fields are named incorrectly according to our naming guidelines, e.g. this should be ```_name``` rather than ```_Name```.

File: src/System.Security.Principal.Windows/src/System/Security/Principal/WindowsIdentity.cs [hunk 14]
```diff
@@ -0,0 +1,1014 @@
...
+            }
+        }
+
+        public override bool IsAuthenticated
+        {
+            get
+            {
+                if (_isAuthenticated == -1)
+                {
+                    {
```
Comment (stephentoub) [inline src/System.Security.Principal.Windows/src/System/Security/Principal/WindowsIdentity.cs]: Unnecessary extra set of braces?

File: src/System.Security.Principal.Windows/src/System/Security/Principal/WindowsIdentity.cs [hunk 15]
```diff
@@ -0,0 +1,1014 @@
...
+
+        private static void RunImpersonatedInternal(SafeAccessTokenHandle token, Action action)
+        {
+            bool isImpersonating;
+            int hr;
+            SafeAccessTokenHandle previousToken = GetCurrentToken(TokenAccessLevels.MaximumAllowed, false, out isImpersonating, out hr);
+            if (previousToken == null || previousToken.IsInvalid)
+                throw new SecurityException(Interop.mincore.GetMessage(hr));
+
+            s_currentImpersonatedToken.Value = isImpersonating ? previousToken : null;
```
Comment (stephentoub) [inline src/System.Security.Principal.Windows/src/System/Security/Principal/WindowsIdentity.cs]: This doesn't need to be reset to its original value in a finally?

File: src/System.Security.Principal.Windows/tests/System.Security.Principal.Windows.Tests.csproj [hunk 16]
```diff
@@ -0,0 +1,27 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<Project ToolsVersion="12.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory), dir.props))\dir.props" />  
+  <PropertyGroup>
+    <Configuration Condition=" '$(Configuration)' == '' ">Debug</Configuration>
+    <Platform Condition=" '$(Platform)' == '' ">AnyCPU</Platform>
+    <OutputType>Library</OutputType>
+    <AssemblyName>System.Security.Principal.Windows.Tests</AssemblyName>
+    <RootNamespace>System.Security.Principal.Windows.Tests</RootNamespace>
+    <ProjectGuid>{6C36F3AC-54A1-4021-9F5D-CDEFF7347277}</ProjectGuid>
```
Comment (stephentoub) [inline src/System.Security.Principal.Windows/tests/System.Security.Principal.Windows.Tests.csproj]: Please add an <UnsupportedPlatforms/> attribute here for both Linux and OSX... these tests should be Windows only and will fail if run elsewhere.

File: src/System.Security.Principal.Windows/tests/WindowsIdentityTests.cs [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+        Assert.NotNull(windowsIdentity);
+        Assert.True(windowsIdentity.IsAnonymous);
+        Assert.False(windowsIdentity.IsAuthenticated);
+    }
+
+    [Fact]
+    public static void ConstructorsAndProperties()
+    {
+        // Retrieve the Windows account token for the current user.
+        IntPtr logonToken = WindowsIdentity.GetCurrent().AccessToken.DangerousGetHandle();
```
Comment (stephentoub) [inline src/System.Security.Principal.Windows/tests/WindowsIdentityTests.cs]: Could this not be problematic / prone to race conditions if the safe handle gets finalized concurrent with usage of logonToken (no AddRef/Releasing is being done to prevent that)?

File: src/System.Security.Principal.Windows/tests/WindowsIdentityTests.cs [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+
+        // Construct a WindowsIdentity object using the input account token.
+        WindowsIdentity windowsIdentity = new WindowsIdentity(logonToken);
```
Comment (stephentoub) [inline src/System.Security.Principal.Windows/tests/WindowsIdentityTests.cs]: In all of these tests, should these identities be disposed of?

File: src/Common/src/Interop/Windows/mincore/Interop.IsEqualDomainSid.cs (hunk 3 above)
Comment (MattGal) [inline src/Common/src/Interop/Windows/mincore/Interop.IsEqualDomainSid.cs]: I went from Security to SecurityBase because "api-ms-win-security-base-l1-1-0.dll" is already in the OSS tree with this name.  Whoever made that decision is a great person to ask... I do not really know but it would involve either duplication or touching lots of other files to fix.

File: src/System.Security.Principal.Windows/tests/WindowsPrincipalTests.cs [hunk 19]
```diff
@@ -0,0 +1,14 @@
...
+using Xunit;
+
+public class WindowsPrincipalTests
+{
+    [Fact]
+    public static void WindowsPrincipalIsInRoleNeg()
+    {
+        WindowsIdentity windowsIdentity = WindowsIdentity.GetAnonymous();
+        WindowsPrincipal windowsPrincipal = new WindowsPrincipal(windowsIdentity);
+        var ret = windowsPrincipal.IsInRole("REDMOND\\nonexist");
```
Comment (stephentoub) [inline src/System.Security.Principal.Windows/tests/WindowsPrincipalTests.cs]: Maybe use some dummy domain name rather than something Microsoft-specific?

File: src/System.Security.Principal/System.Security.Principal.sln [hunk 20]
```diff
@@ -0,0 +1,22 @@
+﻿
+Microsoft Visual Studio Solution File, Format Version 12.00
+# Visual Studio 14
+VisualStudioVersion = 14.0.22823.1
+MinimumVisualStudioVersion = 10.0.40219.1
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "System.Security.Principal", "src\System.Security.Principal.csproj", "{FBE16BC8-AE2D-422C-861E-861814F53AF7}"
```
Comment (stephentoub) [inline src/System.Security.Principal/System.Security.Principal.sln]: Even though there aren't currently any tests, should we add a test project to make it easy for others to contribute tests?

File: src/Common/src/Interop/Windows/mincore/Interop.IsEqualDomainSid.cs (hunk 3 above)
Comment (stephentoub) [inline src/Common/src/Interop/Windows/mincore/Interop.IsEqualDomainSid.cs]: I was actually referring to the "IsEqualDomainSid" vs "EqualDomainSid" difference.

Comment (MattGal) [inline src/Common/src/Interop/Windows/mincore/Interop.IsEqualDomainSid.cs]: Oh right .  It was initially authored this way, I would guess for readability to make it obvious the output was a bool.  @pallavit  might have some suggestions.

File: src/System.Security.Principal.Windows/src/System.Security.Principal.Windows.csproj (hunk 9 above)
Comment (MattGal) [inline src/System.Security.Principal.Windows/src/System.Security.Principal.Windows.csproj]: I preferred the P2P because if someone were to change Sys.Windows.Principal in a subtle way (especially given it lacks tests right now), we might miss it in testing.  I'm happy to change it the other way. Pretty sure I've seen examples going both ways although primarily Nuget .

File: src/Common/src/Interop/Windows/NtDll/Interop.NtStatus.cs (hunk 2 above)
Comment (MattGal) [inline src/Common/src/Interop/Windows/NtDll/Interop.NtStatus.cs]: should be addressed, nit away if you see more.

File: src/System.Security.Principal.Windows/src/System/Security/Principal/IRCollection.cs (hunk 11 above)
Comment (MattGal) [inline src/System.Security.Principal.Windows/src/System/Security/Principal/IRCollection.cs]: sure, done.

File: src/System.Security.Principal.Windows/tests/System.Security.Principal.Windows.Tests.csproj (hunk 16 above)
Comment (MattGal) [inline src/System.Security.Principal.Windows/tests/System.Security.Principal.Windows.Tests.csproj]: done below.

**Comment by MattGal** [comment]
@dotnet-bot test this please

File: src/Common/src/Interop/Windows/mincore/Interop.IsEqualDomainSid.cs (hunk 3 above)
Comment (pallavit) [inline src/Common/src/Interop/Windows/mincore/Interop.IsEqualDomainSid.cs]: This is how it was initially authored before I moved the code to the common location. We should be able to rename it though.

**Comment by stephentoub** [comment]
Can you open an issue to track the feedback that wasn't addressed?  Some of those look like they could be legitimate bugs, or at least deserve comments as to why they're not.

Otherwise, LGTM.

**Comment by Priya91** [comment]
LGTM
